### PR TITLE
Avoid stack overflow by reducing stack usage of `BinaryExpr::evaluate` in debug builds

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Closes https://github.com/apache/arrow-datafusion/issues/419

 # Rationale for this change
Prior to this PR, trying to evaluate `BinaryExprs` more than a few level deep (e.g. something like `a + a + a + a + a + a + a + a ...`) would result in a stack overflow

For example the test included  fails like this with `tree_depth` of 10 (in debug builds)

```
thread 'physical_plan::expressions::binary::tests::relatively_deeply_nested' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass '-p datafusion --lib'

Caused by:
  process didn't exit successfully: `/Users/alamb/Software/arrow-datafusion/target/debug/deps/datafusion-68280be86fef135e deeply` (signal: 6, SIGABRT: process abort signal)
```

# What changes are included in this PR?
1. Break the `BinaryExpr::evaluate` into a few smaller functions
2. Remove special case workaround added for #910 

# Are there any user-facing changes?
Not really other than avoiding stack overflows while evaluating queries in debug builds

# Technical Backstory

I believe the issue is that in debug builds, each local variable gets its own (unique space in the stack). Due to the size of `BinaryExpr::evaluate` (largely hidden by macros) this results in a ludicrous amount of stack required for each call to `BinaryExpr::evaluate`. Since `BinaryExpr::evaluate` is implemented recursively this means even a few nesting levels exhausts a 2MB stack. 

You can see evidence of  looking at the disassembly. Here is how I did it on a mac:

```shell
otool -vt target/debug/deps/datafusion-68280be86fef135e > /tmp/df.asm
```

And the associated assembly shows the stack size to be 0x55a10 (350736 bytes)
```asm
__ZN118_$LT$datafusion..physical_plan..expressions..binary..BinaryExpr$u20$as$u20$datafusion..physical_plan..PhysicalExpr$GT$8evaluate17h2877dfaf102fa64eE:
0000000100d316d0	pushq	%rbp
0000000100d316d1	movq	%rsp, %rbp
0000000100d316d4	movl	$0x55a10, %eax                  ## the subq instruction below uses this value
0000000100d316d9	callq	0x1025dc540                     ## to effectively add 350736 (350K!) to the stack pointer
0000000100d316de	subq	%rax, %rsp
0000000100d316e1	movq	%rdx, -0x4a998(%rbp)
0000000100d316e8	movq	%rsi, -0x4a990(%rbp)
0000000100d316ef	movq	%rdi, %rax
...
```

In case you were curious, the same function in a  release build only uses 0x398 (920 bytes) of stack space.

```asm
__ZN118_$LT$datafusion..physical_plan..expressions..binary..BinaryExpr$u20$as$u20$datafusion..physical_plan..PhysicalExpr$GT$8evaluate17h51f0595db0e2c70cE:
00000001009468a0        pushq   %rbp
00000001009468a1        movq    %rsp, %rbp
00000001009468a4        pushq   %r15
00000001009468a6        pushq   %r14
00000001009468a8        pushq   %r13
00000001009468aa        pushq   %r12
00000001009468ac        pushq   %rbx
00000001009468ad        subq    $0x398, %rsp                    ## imm = 0x398
```

For those of you visually minded, here is an illustration:

```text
┌ ─ ─ ─ ─ ─ ─ ─ ─ ─
                   │
│     Function                      Base Pointer
     Parameters    │     ┌──────       (%ebp)
│                        │
┌──────────────────┤     │
│  Return Address  │     │
├──────────────────┤     │
│Saved Base Pointer│◀────┘
├──────────────────┘
    Local Var 1    │
├ ─ ─ ─ ─ ─ ─ ─ ─ ─
    Local Var 2    │             I theorize that in a debug build, a
├ ─ ─ ─ ─ ─ ─ ─ ─ ─           distinct space is reserved for each local
        ...        │                  variable of the function.
├ ─ ─ ─ ─ ─ ─ ─ ─ ─
    Local Var N    │            Thus a large number of local variables
├ ─ ─ ─ ─ ─ ─ ─ ─ ─            will result in a large stack frame (and
                   │              thus a large amount of stack space
│     Function                         consumed for each call)
     Parameters    │
│
┌──────────────────┤
│  Return Address  │◀────┐
└──────────────────┘     │
                         │
                         │
                         │         Stack Pointer
                         └──────       (%esp)
```

The "Fix" if you will is to break BinaryExpr into smaller functions that each require less stack space

The stack frame size after this PR is 0x770 = 1904 bytes

```asm
__ZN118_$LT$datafusion..physical_plan..expressions..binary..BinaryExpr$u20$as$u20$datafusion..physical_plan..PhysicalExpr$GT$8evaluate17h2877dfaf102fa64eE:
0000000100d31a50        pushq   %rbp
0000000100d31a51        movq    %rsp, %rbp
0000000100d31a54        subq    $0x770, %rsp                    ## imm = 0x770
0000000100d31a5b        movq    %rdx, -0x6a8(%rbp)
0000000100d31a62        movq    %rsi, -0x6a0(%rbp)
```

Note that even though some of the new functions require non trivial stack size (listed below), a major difference is they are not called recursively and thus there is only ever one frame of them on the stack:

```
evaluate_array_scalar:       0x1DE80 (122496 bytes)
evaluate_scalar_array:       0x162A0 (90784 bytes)
evaluate_with_resolved_args: 0x1F970 (129392 bytes)
```

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -105,8 +105,6 @@ jobs:
         run: |
           export ARROW_TEST_DATA=$(pwd)/testing/data
           export PARQUET_TEST_DATA=$(pwd)/parquet-testing/data
-          # run tests on all workspace members with default feature list + avro
```
Comment (alamb) [inline .github/workflows/rust.yml:108]: This is the workaround added in #910

File: datafusion/src/physical_plan/expressions/binary.rs [hunk 2]
```diff
@@ -543,86 +543,17 @@ impl PhysicalExpr for BinaryExpr {
             )));
         }
 
+        // Attempt to use special kernels if one input is scalar and the other is an array
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/binary.rs:546]: There are no intended changes to this function's behavior, simply breaking it up into several smaller functions

File: datafusion/src/physical_plan/expressions/binary.rs [hunk 3]
```diff
@@ -1381,4 +1416,37 @@ mod tests {
...
+        // the stack in debug builds
+
+        let input: Vec<_> = vec![1, 2, 3, 4, 5].into_iter().map(Some).collect();
+        let a: Int32Array = input.iter().collect();
+
+        let batch = RecordBatch::try_from_iter(vec![("a", Arc::new(a) as _)]).unwrap();
+        let schema = batch.schema();
+
+        // build a left deep tree ((((a + a) + a) + a ....
+        let tree_depth: i32 = 100;
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/binary.rs:1434]: On master, this test causes a stack overflow with `tree_depth` of `10`. After the changes in this PR it passes successfully with a tree_depth of 100 (I didn't try any bigger)

**Review by Dandandan — APPROVED** [review]

File: datafusion/src/physical_plan/expressions/binary.rs (hunk 3 above)
Comment (NGA-TRAN) [reply datafusion/src/physical_plan/expressions/binary.rs:1434]: 🎉

**Review by NGA-TRAN — APPROVED** [review]
Thanks for the deep-dive evaluation and the fix, @alamb

**Comment by alamb** [comment]
I'll plan to merge this tomorrow unless there are objections

**Review by houqp — APPROVED** [review]
Great job on the deep dive and detailed diagram :)

**Comment by houqp** [comment]
This kind of "match to many arms with macros" pattern is very common in our code base. It seems like in the future if we want to get fancy, we could automate detection of problems like this in our code base. For example, having a linter tool automatically find out all recursive functions and check their stack size to see if it has passed a certain threshold.

**Comment by alamb** [comment]
> This kind of "match to many arms with macros" pattern is very common in our code base. It seems like in the future if we want to get fancy, we could automate detection of such problem in our code base. For example, having a linter tool automatically find out all recursive functions and check their stack size to see if it has passed a certain threshold.

I agree -- if we run into this problem again, investing in an automated tool like that sounds like a good idea.

**Pull request merged**